packages/dht/src/dht/PeerManager.ts

```diff
@@ -0,0 +1,320 @@
+        )
+        this.randomPeers.on('newContact', (newContact: RemoteDhtNode, closestContacts: RemoteDhtNode[]) =>
+            this.emit('newRandomContact', newContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+    }
+
+    private onKBucketPing(oldContacts: RemoteDhtNode[], newContact: RemoteDhtNode): void {
+        if (this.stopped) {
+            return
+        }
+        const sortingList: SortedContactList<RemoteDhtNode> = new SortedContactList(this.config.ownPeerId!, 100)
+        sortingList.addContacts(oldContacts)
+        const sortedContacts = sortingList.getAllContacts()
+        this.config.connectionManager?.weakUnlockConnection(sortedContacts[sortedContacts.length - 1].getPeerDescriptor())
+        this.bucket?.remove(sortedContacts[sortedContacts.length - 1].getPeerId().value)
+        this.bucket!.add(newContact)
+    }
+
+    private onKBucketRemoved(contact: RemoteDhtNode): void {
+        if (this.stopped) {
+            return
+        }
+        this.config.connectionManager?.weakUnlockConnection(contact.getPeerDescriptor())
+        logger.trace(`Removed contact ${contact.getPeerId().value.toString()}`)
+        this.emit(
+            'kbucketContactRemoved',
+            contact.getPeerDescriptor()
+        )
+        if (this.bucket!.count() === 0) {
+            this.emit('kBucketEmpty')
+        }
+    }
+
+    private onKBucketAdded(contact: RemoteDhtNode): void {
+        if (this.stopped) {
+            return
+        }
+        if (!this.stopped && !contact.getPeerId().equals(this.config.ownPeerId!)) {
+            // Important to lock here, before the ping result is known
+            this.config.connectionManager?.weakLockConnection(contact.getPeerDescriptor())
+            if (this.connections.has(contact.getPeerId().toKey())) {
+                logger.trace(`Added new contact ${contact.getPeerId().value.toString()}`)
+                this.emit(
+                    'newKbucketContact',
+                    contact.getPeerDescriptor(),
+                    this.neighborList!.getClosestContacts(this.config.getClosestContactsLimit).map((peer) => peer.getPeerDescriptor())
+                )
+            } else {    // open connection by pinging
+               
+                contact.ping().then((result) => {
+                    if (result) {
+                        logger.trace(`Added new contact ${contact.getPeerId().value.toString()}`)
+                        this.emit(
+                            'newKbucketContact',
+                            contact.getPeerDescriptor(),
+                            this.neighborList!.getClosestContacts(this.config.getClosestContactsLimit).map((peer) => peer.getPeerDescriptor())
+                        )
+                    } else {
+                        this.config.connectionManager?.weakUnlockConnection(contact.getPeerDescriptor())
```

Comment:
It is never `undefined`

```suggestion
                        this.config.connectionManager.weakUnlockConnection(contact.getPeerDescriptor())
```
(also 4 other places in this file)